lib/std/math/isnormal.zig

```diff
@@ -1,45 +1,48 @@
 const std = @import("../std.zig");
 const math = std.math;
 const expect = std.testing.expect;
-const maxInt = std.math.maxInt;
 
-// Returns whether x has a normalized representation (i.e. integer part of mantissa is 1).
+/// Returns whether x is neither zero, subnormal, infinity, or NaN.
 pub fn isNormal(x: anytype) bool {
     const T = @TypeOf(x);
-    switch (T) {
-        f16 => {
-            const bits = @bitCast(u16, x);
-            return (bits + (1 << 10)) & (maxInt(u16) >> 1) >= (1 << 11);
-        },
-        f32 => {
-            const bits = @bitCast(u32, x);
-            return (bits + (1 << 23)) & (maxInt(u32) >> 1) >= (1 << 24);
-        },
-        f64 => {
-            const bits = @bitCast(u64, x);
-            return (bits + (1 << 52)) & (maxInt(u64) >> 1) >= (1 << 53);
-        },
-        f128 => {
-            const bits = @bitCast(u128, x);
-            return (bits + (1 << 112)) & (maxInt(u128) >> 1) >= (1 << 113);
-        },
-        else => {
-            @compileError("isNormal not implemented for " ++ @typeName(T));
-        },
+    const TBits = std.meta.Int(.unsigned, @bitSizeOf(T));
+    if (@typeInfo(T) != .Float) {
+        @compileError("isNormal not implemented for " ++ @typeName(T));
     }
+
+    const increment_exp = 1 << math.floatMantissaBits(T);
+    const remove_sign = ~@as(TBits, 0) >> 1;
+
+    // Hack: 1 is added to the exponent, and if it overflows to 0 or becomes 1,
+    // then it was all zeroes (subnormal) or all ones (non-finite, like inf/nan).
+    // The sign bit is removed because all ones would overflow into it,
+    // as well as the sign bit in general getting in the way.
+    const float_bits = @bitCast(TBits, x);
+    const result = float_bits +% increment_exp;
+
+    return result & remove_sign >= (increment_exp << 1);
 }
 
 test "math.isNormal" {
-    try expect(!isNormal(math.nan(f16)));
-    try expect(!isNormal(math.nan(f32)));
-    try expect(!isNormal(math.nan(f64)));
-    try expect(!isNormal(math.nan(f128)));
-    try expect(!isNormal(@as(f16, 0)));
-    try expect(!isNormal(@as(f32, 0)));
-    try expect(!isNormal(@as(f64, 0)));
-    try expect(!isNormal(@as(f128, 0)));
-    try expect(isNormal(@as(f16, 1.0)));
-    try expect(isNormal(@as(f32, 1.0)));
-    try expect(isNormal(@as(f64, 1.0)));
-    try expect(isNormal(@as(f128, 1.0)));
+    inline for ([_]type{ f16, f32, f64, f128 }) |T| {
+        const TBits = std.meta.Int(.unsigned, @bitSizeOf(T));
+        const mantissa_bits = math.floatMantissaBits(T);
+
+        // normals
+        try expect(isNormal(@as(T, 1.0)));
+
+        // subnormals
+        try expect(!isNormal(@as(T, -0.0)));
+        try expect(!isNormal(@as(T, 0.0)));
+        try expect(!isNormal(@bitCast(T, @as(TBits, 1))));
```

Comment:
This is just the 'true min', e.g. `f128_true_min` - maybe preferable to use the variable name/function to make this clearer? `floatTrueMin(T)`

Reply:
I don't think there *is* a `floatTrueMin`(?), but adding one sounds like a good idea.

Reply:
Indeed, I was suggesting adding, as in my other comment :)
> Maybe add the max and min normals, e.g. `f128_max` and `f128_min` (add a function that takes the float type in `math.zig`?).

Reply:
Apologies for the delay, I saw that first thing when I woke up and had to leave. I understand what you mean now. Here's a thought - right now float mins and maxes are just done as magic constants, which isn't very Zig-y and this PR is kind of trying to fix that. How about adding a new file, `math/float.zig`, where all these constant derivation functions can exist instead of being scattered throughout `math.zig` and individual files like `math/epsilon.zig`, `math/inf.zig`, etc., and have the existing constants like `f32_max` be defined as for example `floatMax(f32)` (since replacing usages of it is **really** outside the scope of this PR)? This file would also be a logical place for the aforementioned test cases to exist.

Reply:
Yep, big fan of that plan :) Fine if you think the change should be done in a separate PR (or I'd be happy to do it).

I put together https://github.com/LewisGaul/zig/wiki/Maths-APIs today, where these suggested functions are mentioned. Note the existence of `maxInt()` and `minInt()`, which should maybe be renamed to `intMax()` and `intMin()` at some point to be consistent with the float equivalents (where I don't think `trueMinFloat()` would be a great choice of name).

I think all of `inf()`, `nan()`, `snan()`, `floatMax()`, `floatMin()`, `floatTrueMin()`, `floatEps()` should be implemented generically in terms of `floatMantissaBits()` and `floatExponentBits()` (where required) or just simpler bit manipulations. Then it would make sense for the constants to be defined using the functions IMO (and wouldn't suggest removing the constants).

Reply:
Glad we're on the same wavelength, I had the exact same thing where I tried to match `maxInt` and indeed ended up with `trueMinFloat` and decided against it hahaha. I was thinking it should be in terms of exponent and mantissa bits too and I like your proposal, the only thing that sticks out to me is the `floatToint` constant which is a name borrowed from C code, meaning "to int" (it is 1/ε).

Other than that, I'd be glad to rebase the PR as soon as I get home, if that sounds alright.